src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/BuildHostProcessManager.cs

```diff
@@ -0,0 +1,238 @@
+    {
+        var neededBuildHostKind = GetKindForProject(projectFilePath);
+
+        using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+        {
+            if (!_processes.TryGetValue(neededBuildHostKind, out var buildHostProcess))
+            {
+                var process = neededBuildHostKind switch
+                {
+                    BuildHostProcessKind.NetCore => LaunchDotNetCoreBuildHost(),
+                    BuildHostProcessKind.NetFramework => LaunchDotNetFrameworkBuildHost(),
+                    _ => throw ExceptionUtilities.UnexpectedValue(neededBuildHostKind)
+                };
+
+                buildHostProcess = new BuildHostProcess(process);
+                buildHostProcess.Disconnected += BuildHostProcess_Disconnected;
+                _processes.Add(neededBuildHostKind, buildHostProcess);
+            }
+
+            return buildHostProcess.BuildHost;
+        }
+    }
+
+    private void BuildHostProcess_Disconnected(object? sender, EventArgs e)
+    {
+        Contract.ThrowIfNull(sender, $"{nameof(BuildHostProcess)}.{nameof(BuildHostProcess.Disconnected)} was raised with a null sender.");
+
+        Task.Run(async () =>
+        {
+            BuildHostProcess? processToDispose = null;
+
+            using (await _gate.DisposableWaitAsync().ConfigureAwait(false))
+            {
+                // Remove it from our map; it's possible it might have already been removed if we had more than one way we observed a disconnect.
+                var existingProcess = _processes.SingleOrNull(p => p.Value == sender);
+                if (existingProcess.HasValue)
+                {
+                    processToDispose = existingProcess.Value.Value;
+                    _processes.Remove(existingProcess.Value.Key);
+                }
+            }
+
+            // Dispose outside of the lock (even though we don't expect much to happen at this point)
+            if (processToDispose != null)
+                await processToDispose.DisposeAsync();
+        });
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        foreach (var process in _processes.Values)
+            await process.DisposeAsync();
+    }
+
+    private Process LaunchDotNetCoreBuildHost()
+    {
+        var processStartInfo = new ProcessStartInfo()
+        {
+            FileName = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "dotnet.exe" : "dotnet",
+            CreateNoWindow = true,
+            UseShellExecute = false,
+            RedirectStandardInput = true,
+            RedirectStandardOutput = true,
+        };
+
+        // We need to roll forward to the latest runtime, since the project may be using an SDK (or an SDK required runtime) newer than we ourselves built with.
+        // We set the environment variable since --roll-forward LatestMajor doesn't roll forward to prerelease SDKs otherwise.
+        processStartInfo.Environment["DOTNET_ROLL_FORWARD_TO_PRERELEASE"] = "1";
```

Comment:
Good questions!

> So what we're doing here is launching the build host with whatever the latest major version they have installed is right?

At the moment, yes, since MSBuildLocator currently requires that you have to be running the same or higher major version to even discover the applicable SDK in the first place; it's an annoying catch-22. And once that process is launched, we will find the global.json for the first project we then ask it to process when we invoke MSBuildLocator there.

Once this is in, I plan on making a change to MSBuildLocator that would allow us to discover (but not register) the applicable SDK version directly. In that case:

1. We'll know the required runtime version of the build process, so we can avoid launching a prerelease runtime for a released SDK.
2. We'll also be able to have different build host processes for different SDK versions, for the more esoteric cases of a huge monorepo with different global.jsons. DevKit doesn't technically support that scenario (and we don't either today), but it's something that will fall out in the wash automatically.

> What directory are we launching this from

This is still a reminder for me though that was brought up that DevKit might explicitly set a working directory so there's no risk this locks a user folder underneath us.